shared_client.go

```diff
@@ -0,0 +1,236 @@
+	if key == "" {
+		// Simplified case when the client is actually not shared
+		return newSharedClient(conf, serverAddrStr), func() {
+			close(client.requestsToSend)
+			client.conn = nil
+			client.wg.Wait()
+		}
+	}
+	for {
+		// lock for s.clients access
+		s.lock.Lock()
+		// locate client to re-use if possible.
+		client = s.clients[key]
+		if client == nil {
+			client = newSharedClient(conf, serverAddrStr)
+			s.clients[key] = client
+			s.lock.Unlock()
+			// new client, we are done
+			break
+		}
+		s.lock.Unlock()
+
+		// reusing client that may start closing while we wait for its lock
+		client.Lock()
+		if client.refcount > 0 {
+			// not closed, add our refcount
+			client.refcount++
+			client.Unlock()
+			break
+		}
+		// client was closed while we waited for it's lock, discard and try again
+		client.Unlock()
+		client = nil
+	}
+
+	return client, func() {
+		client.Lock()
+		defer client.Unlock()
+		client.refcount--
+		if client.refcount == 0 {
+			// connection close must be completed while holding the client's lock to
+			// avoid a race where a new client dials using the same 5-tuple and gets a
+			// bind error.
+			// The client remains findable so that new users with the same key may wait
+			// for this closing to be done with.
+			close(client.requestsToSend)
+			client.conn = nil
+			client.wg.Wait()
+			// Make client unreachable
+			// Must take s.lock for this.
+			s.lock.Lock()
+			if key != "" {
+				delete(s.clients, key)
+			}
+			s.lock.Unlock()
+		}
+	}
+}
+
+type request struct {
+	ctx context.Context
+	msg *Msg
+	ch  chan sharedClientResponse
+}
+
+type sharedClientResponse struct {
+	msg *Msg
+	rtt time.Duration
+	err error
+}
+
+// A SharedClient keeps state for concurrent transactions on the same upstream client/connection.
+type SharedClient struct {
+	serverAddr string
+
+	*Client
+
+	requestsToSend chan request
+	responses      chan sharedClientResponse
+
+	wg sync.WaitGroup
+
+	sync.Mutex // protects the fields below
+	refcount   int
+	conn       *Conn
+}
+
+func newSharedClient(conf *Client, serverAddr string) *SharedClient {
+	return &SharedClient{
+		refcount:       1,
+		serverAddr:     serverAddr,
+		Client:         conf,
+		requestsToSend: make(chan request, 1),
+		responses:      make(chan sharedClientResponse, 1),
+	}
+}
+
+// ExchangeShared dials a connection to the server on first invocation, and starts a handler
+// goroutines to send and receive responses, distributing them to appropriate concurrent caller
+// based on the DNS message Id.
+func (c *SharedClient) ExchangeShared(m *Msg) (r *Msg, rtt time.Duration, err error) {
+	return c.ExchangeSharedContext(context.Background(), m)
+}
+
+// handler is started when the connection is dialed
+func handler(wg *sync.WaitGroup, client *Client, conn *Conn, requests chan request, responses chan sharedClientResponse) {
+	defer wg.Done()
+
+	// Receive loop
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
```

Comment:
I think we can get rid of these `wg.Add` and `wg.Done()` here. That's because `handler` can't exit until `responses` is closed, and only after that the `wg.Done()` deferred at the beginning of the function will be called.

Reply:
It seems to me that handler can exit before `responses` is closed, for example if `requests` is closed. I saw this happen in tests when I was working on getting rid of channel buffers (which can hide bugs).

Reply:
You're right, I've missed that case. :+1: